contracts/multisig/contracts/src/AssetProxyOwner.sol

```diff
@@ -16,93 +16,205 @@
+            uint256[] memory values
+        ) = abi.decode(
+            transaction.data,
+            (bytes[], address[], uint256[])
+        );
+        // solhint-enable
+
+        // Ensure lengths of array properties are equal
+        uint256 length = data.length;
+        require(
+            length == destinations.length && length == values.length,
+            "EQUAL_LENGTHS_REQUIRED"
+        );
+
+        uint256 transactionConfirmationTime = confirmationTimes[transactionId];
+        for (uint i = 0; i != length; i++) {
+            // Ensure that each function call is past its timelock
+            _assertValidFunctionCall(
+                transactionConfirmationTime,
+                data[i],
+                destinations[i]
+            );
+            // Call each function
+            // solhint-disable-next-line avoid-call-value
+            (bool didSucceed,) = destinations[i].call.value(values[i])(data[i]);
+            // Ensure that function call was successful
+            require(
+                didSucceed,
+                "FAILED_EXECUTION"
+            );
+        }
+        emit Execution(transactionId);
+    }
+
+    /// @dev Registers a custom timelock to a specific function selector / destination combo
+    /// @param hasCustomTimeLock True if timelock is custom.
+    /// @param functionSelector 4 byte selector of registered function.
+    /// @param destination Address of destination where function will be called.
+    /// @param newSecondsTimeLocked Duration in seconds needed after a transaction is confirmed to become executable.
+    function _registerFunctionCall(
+        bool hasCustomTimeLock,
+        bytes4 functionSelector,
+        address destination,
+        uint128 newSecondsTimeLocked    
+    )
+        internal
+    {
+        // Clear the previous secondsTimeLocked if custom timelock not used
+        uint128 _secondsTimeLocked = hasCustomTimeLock ? newSecondsTimeLocked : 0;
+        TimeLock memory timeLock = TimeLock({
+            hasCustomTimeLock: hasCustomTimeLock,
+            secondsTimeLocked: _secondsTimeLocked
+        });
+        functionCallTimeLocks[functionSelector][destination] = timeLock;
+        emit FunctionCallTimeLockRegistration(
+            functionSelector,
+            destination,
+            hasCustomTimeLock,
+            _secondsTimeLocked
+        );
+    }
+
+    /// @dev Ensures that the function call has past its timelock.
+    /// @param transactionConfirmationTime Timestamp at which transaction was fully confirmed.
+    /// @param data Function calldata.
+    /// @param destination Address to call function on.
+    function _assertValidFunctionCall(
+        uint256 transactionConfirmationTime,
+        bytes memory data,
+        address destination
+    )
+        internal
+        view
+    {
+        bytes4 functionSelector = data.readBytes4(0);
+        TimeLock storage timeLock = functionCallTimeLocks[functionSelector][destination];
```

Comment:
Isn't `timeLock` just a pointer to the storage slot here and not actually using an `sload`? Either way, I tested it and using `memory` is a bit cheaper. Went ahead and made the change.

Reply:
That's right, although with a storage pointer every time you access a field it'll do an `sload`. Whereas with `memory` it'll at most do an `sload` on the first access; and it _should_ only use a single `sload` in this assignment since the entire struct fits into 1 word (although I'm not positive Solidity is that clever).